File: src/2025-02-24-deepclaude-rust/java/exercises/practice/book-store/src/main/java/BookStore.java
import java.util.*;
import java.util.stream.Collectors;

class BookStore {

    private static final double[] DISCOUNTS = {0, 0, 0.05, 0.10, 0.20, 0.25};
    private Map<String, Double> memo;

    double calculateBasketCost(List<Integer> books) {
        // Count occurrences of each book (1-5)
        int[] countsArray = new int[5];
        for (int book : books) {
            countsArray[book - 1]++;
        }

        // Convert to sorted list in descending order, removing zeros
        List<Integer> counts = Arrays.stream(countsArray)
                .boxed()
                .sorted(Collections.reverseOrder())
                .filter(c -> c > 0)
                .collect(Collectors.toList());

        memo = new HashMap<>();
        return calculateCost(counts);
    }

    private double calculateCost(List<Integer> counts) {
        if (counts.isEmpty()) {
            return 0.0;
        }

        String key = String.join(",", counts.stream().map(String::valueOf).collect(Collectors.toList()));
        if (memo.containsKey(key)) {
            return memo.get(key);
        }

        double minTotal = Double.MAX_VALUE;

        int maxSetSize = counts.size();
        for (int k = 1; k <= maxSetSize; k++) {
            List<Integer> nextCounts = new ArrayList<>(counts);
            // Subtract 1 from first k elements
            for (int i = 0; i < k; i++) {
                nextCounts.set(i, nextCounts.get(i) - 1);
            }

            // Filter zeros, sort descending, and recurse
            List<Integer> newCounts = nextCounts.stream()
                    .filter(c -> c > 0)
                    .sorted(Collections.reverseOrder())
                    .collect(Collectors.toList());

            double setCost = k * 8 * (1 - DISCOUNTS[k]);
            double remainingCost = calculateCost(newCounts);
            double total = setCost + remainingCost;

            if (total < minTotal) {
                minTotal = total;
            }
        }

        memo.put(key, minTotal);
        return minTotal;
    }
}

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/book-store/book_store.go
package bookstore

import (
	"sort"
)

var groupCost = [6]int{0, 800, 1520, 2160, 2560, 3000}

func Cost(books []int) int {
	counts := make(map[int]int)
	for _, book := range books {
		counts[book]++
	}

	var list []int
	for _, c := range counts {
		list = append(list, c)
	}

	// Pad with zeros to ensure exactly 5 elements
	for len(list) < 5 {
		list = append(list, 0)
	}

	// Sort in descending order
	sort.Slice(list, func(i, j int) bool {
		return list[i] > list[j]
	})

	var state [5]int
	copy(state[:], list)

	memo := make(map[[5]int]int)
	return calculate(state, memo)
}

func calculate(state [5]int, memo map[[5]int]int) int {
	if val, ok := memo[state]; ok {
		return val
	}

	sum := 0
	for _, c := range state {
		sum += c
	}
	if sum == 0 {
		return 0
	}

	minCost := -1
	for k := 5; k >= 1; k-- {
		// Check if we can form a group of size k
		valid := true
		for i := 0; i < k; i++ {
			if state[i] == 0 {
				valid = false
				break
			}
		}

		if valid {
			newState := state
			// Subtract 1 from each of the first k elements
			for i := 0; i < k; i++ {
				newState[i]--
			}
			// Sort the new state
			sort.Slice(newState[:], func(i, j int) bool {
				return newState[i] > newState[j]
			})

			cost := groupCost[k] + calculate(newState, memo)
			if minCost == -1 || cost < minCost {
				minCost = cost
			}
		}
	}

	memo[state] = minCost
	return minCost
}
